src/alert.cpp

```diff
@@ -2105,4 +2107,62 @@ namespace libtorrent {
 		return buf;
 	}
 
+	dht_live_nodes_alert::dht_live_nodes_alert(aux::stack_allocator& alloc
+		, sha1_hash const& nid
+		, std::vector<std::pair<sha1_hash, udp::endpoint>> const& nodes)
+		: node_id(nid)
+		, m_alloc(alloc)
+		, m_num_nodes(int(nodes.size()))
+	{
+		std::size_t total_size = nodes.size(); // num bytes for sizes
+		for (auto const& n : nodes) {
+			total_size += n.first.size() + n.second.size();
+		}
+
+		m_nodes_idx = alloc.allocate(aux::numeric_cast<int>(total_size));
+
+		char *ptr = alloc.ptr(m_nodes_idx);
+		for (auto const& n : nodes) {
+			std::size_t size = n.first.size();
+			std::memcpy(ptr, n.first.data(), size);
+			ptr += size;
+			size = n.second.size();
+			TORRENT_ASSERT(size < 0x100);
+			detail::write_uint8(size, ptr);
+			std::memcpy(ptr, n.second.data(), size);
+			ptr += size;
+		}
+	}
+
+	std::string dht_live_nodes_alert::message() const
+	{
+		char msg[200];
+		std::snprintf(msg, sizeof(msg), "dht live nodes for id: %s, nodes %d"
+			, aux::to_hex(node_id).c_str(), m_num_nodes);
+		return msg;
+	}
+
+	int dht_live_nodes_alert::num_nodes() const
+	{
+		return m_num_nodes;
+	}
+
+	std::vector<std::pair<sha1_hash, udp::endpoint>> dht_live_nodes_alert::nodes() const
+	{
+		std::size_t const num_nodes = aux::numeric_cast<std::size_t>(m_num_nodes);
+		std::vector<std::pair<sha1_hash, udp::endpoint>> nodes(num_nodes);
```

Comment:
I think it would be a bit nicer to use ``reserve()`` here and ``emplace_back()`` in the loop. That way you may be able to cut down on the ``memcpy()`` too. actually, I think it would be beneficial to store the node endpoints in a more compact format. i.e. 4-16 bytes for the IP address and 2 bytes for the port. The ``sockaddr_in`` and ``sockaddr_in6`` are likely a lot larger than that